Link to address literals from address value types (#12954)
ethereum/solidity: docs/types/value-types.rst

    declare its type as ``address payable`` to make this requirement visible. Also,
    try to make this distinction or conversion as early as possible.

    The distinction between ``address`` and ``address payable`` was introduced with version 0.5.0.

>     Also starting from that version, contracts do not derive from the address type, but can still be explicitly converted to

Reviewer: I see that this was here before your PR but it seems weird. Was this actually inheritance? I always thought that it was an explicit conversion plus the presence of contract members (which has a similar effect but it's technically not the same thing):

```suggestion
    Also starting from that version, contracts are not implicitly convertible to the ``address`` type, but can still be explicitly converted to
```

    or you can use ``address(uint160(uint256(b)))``, which results in ``0x777788889999AaAAbBbbCcccddDdeeeEfFFfCcCc``.

.. note::

>     See :ref:`Address Literals`, which may be written as mixed-case hexadecimals conforming to `EIP-55 <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md>`_.

Reviewer: This seems like something that could just be integrated into the text instead of keeping it in a note:

```suggestion
Mixed-case hexadecimal numbers conforming to `EIP-55 <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md>`_ are automatically treated as literals of the ``address`` type.
See :ref:`Address Literals`.
```

>     See :ref:`Address Literals`, which may be written as mixed-case hexadecimals conforming to `EIP-55 <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md>`_.

Reviewer: Actually, there's an error here. I think that referencing other headings only works only works within the same document. You probably need to use an explicit anchor.
